zfit/core/minimizer.py

```diff
@@ -4,3 +4,266 @@
 """
 
 from __future__ import print_function, division, absolute_import
+
+import abc
+import collections
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+import zfit.core.math as zmath
+import zfit.core.tfext as ztf
+import zfit.ztf
+
+
+class AbstractMinimizer(object):
+    """Define the minimizer interface."""
+
+    @abc.abstractmethod
+    def minimize(self):
+        raise NotImplementedError
+
+    def _minimize(self, *args, **kwargs):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def edm(self):
+        raise NotImplementedError
+
+    def _edm(self, *args, **kwargs):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def fmin(self):
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def tolerance(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def status(self):
+        raise NotImplementedError
+
+
+class BaseMinimizer(object):
+
+    def __init__(self, name="BaseMinimizer", tolerance=1e-8, sess=None, *args, **kwargs):
+        super(BaseMinimizer, self).__init__(*args, **kwargs)
+        self.name = name
+
+        self.sess = sess
+        self.tolerance = tolerance
+
+    @property
+    def tolerance(self):
+        return self._tolerance
+
+    @staticmethod
+    def gradient_par(func):
+        return zmath.gradient_par(func)
+
+    @staticmethod
+    def start_values(parameters):
+        """Extract the current value if defined, otherwise random.
+
+        Arguments:
+            parameters (FitParameter):
+
+        Return:
+            list(const): the current values of parameters
+        """
+        values = [p.read_value() for p in parameters]
+        # TODO: implement if initial val not given
+        return values
+
+    @tolerance.setter
+    def tolerance(self, tolerance):
+        self._tolerance = tolerance
+
+    def minimize(self, loss, var_list=None):
```

Comment:
Quick question: minimiser doesn't evaluate errors?

Reply:
Not yet, but will. I mostly focused on the basepdf here, the minimizer is a fast-setup thing to have something work together.

As you asked here anyway: A better name for `loss`? `cost`?